src/main/groovy/com/google/protobuf/gradle/plugins/ProtobufPlugin.groovy

```diff
@@ -0,0 +1,86 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.protobuf.gradle.plugins
+
+import org.gradle.api.Action
+import org.gradle.api.GradleException
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.plugins.AppliedPlugin
+
+class ProtobufPlugin implements Plugin<Project> {
+    // any one of these plugins should be sufficient to proceed with applying this plugin
+    private static final List<String> javaPluginOptions = ['java']
+    private static final List<String> androidPluginOptions = ['com.android.application',
+                                                              'com.android.library',
+                                                              'android',
+                                                              'android-library']
+
+    private List<String> pluginsApplied = [];
+    private Project project;
+
+    void apply(final Project project) {
+        this.project = project
+
+        // At least one of the prerequisite plugins must by applied before this plugin can be applied, so
+        // we will use the PluginManager.withPlugin() callback mechanism to delay applying this plugin until
+        // after that has been achieved. If project evaluation completes before one of the prerequisite plugins
+        // has been applied then we will assume that none of prerequisite plugins were specified and we will
+        // throw an Exception to alert the user of this configuration issue.
+        applyWithPrerequisitePlugin(javaPluginOptions, 'com.google.protobuf.java')
```

Comment:
If the plugins for Java and C# were separate, would that mean they will create separate code-generation tasks for each language? Then you may have a problem with the `generateProtoTasks` closure. For example, `ofSourceSet()` currently only selects code-generation tasks for Java. If we allowed Java and C# to be generated at the same time, would `ofSourceSet()` return tasks for both languages? If so, how would we configure the tasks for a specific language?

If everything is under the same plugin, it's still possible to generate for multiple languages, possibly through an additional configuration knob in the `protobuf` block. It won't create additional tasks for additional languages. Instead, options (e.g., `csharp` in addition to `java`) will be added to the `builtins` the tasks, each of which generates for all languages. `ofSourceSet()` will select the task(s) for a source set. Fine-grained control is possible by using the `remove` trick in the `builtins` block. This also means `protoc` is invoked only once for each proto file, instead of twice or more which has performance impact.

Android is special. The code-gen tasks are per-variant, not per-sourceSet, so it probably won't get along with other languages. Probably nobody would want to put Android and another language in the same project.